sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -1612,6 +1612,49 @@ object CodeGenerator extends Logging {
     }
   }
 
+  /**
+   * Extracts all the input variables from references and subexpression elimination states
+   * for a given `expr`. This result will be used to split the generated code of
+   * expressions into multiple functions.
+   */
+  def getLocalInputVariableValues(
+      context: CodegenContext,
+      expr: Expression,
+      subExprs: Map[Expression, SubExprEliminationState]): Seq[((String, DataType), Expression)] = {
+    val argMap = mutable.Map[(String, DataType), Expression]()
+    val stack = mutable.Stack[Expression](expr)
+    while (stack.nonEmpty) {
+      stack.pop() match {
+        case e if subExprs.contains(e) =>
+          val exprCode = subExprs(e)
+          val SubExprEliminationState(isNull, value) = exprCode
+          if (value.isInstanceOf[VariableValue]) {
```

Comment:
When would the `value` / `isNull` be something other than a `VariableValue`? If we don't expect it to be anything else, can we change this to an assertion instead?

Reply:
Yea, that value could be `TrueLiteral`, `FalseLiteral`, or the others. This method needs to collect local variables for split functions because split functions can access global and literal variables directly but they can't do local variables. So, we need to collect local variables here and pass them into split functions.